File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if(nums.length==1) return nums[0];
        int maxResult = 0; 
        int l=1,r=1;
        for(int i=0;i<nums.length;i++){
            l = (l==0)?1:l;
            r = (r==0)?1:r;
            l*=nums[i];
            r*=nums[nums.length-1-i];
            maxResult = Math.max(maxResult, Math.max(l,r));
        }
        return maxResult;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        currMax,currMin = 1,1
        res = max(nums)
        for num in nums:
            if num == 0:
                currMin, currMax = 1,1
                continue
            temp = num * currMax
            currMax = max(num*currMax, num*currMin, num)
            currMin = min(num*currMin, temp, num)
            res = max(res, currMax)
        return res